File: backend/python/app/auth/router.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import Settings, get_settings
from app.database import get_db
from app.dependencies import get_current_user
from app.models import User

from .schemas import (
    ApiKeyResponse,
    LoginRequest,
    RegisterRequest,
    TokenResponse,
    UserResponse,
)
from .service import (
    create_access_token,
    create_refresh_token,
    generate_api_key,
    hash_password,
    verify_password,
)

router = APIRouter(prefix="/auth", tags=["auth"])


@router.post("/register", response_model=TokenResponse, status_code=status.HTTP_201_CREATED)
async def register(
    body: RegisterRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    settings: Annotated[Settings, Depends(get_settings)],
):
    result = await db.execute(select(User).where(User.email == body.email))
    if result.scalar_one_or_none() is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={"error": {"code": "EMAIL_EXISTS", "message": "A user with this email already exists", "details": {}}},
        )

    if len(body.password) < 8:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail={"error": {"code": "WEAK_PASSWORD", "message": "Password must be at least 8 characters", "details": {}}},
        )

    user = User(
        email=body.email,
        password_hash=hash_password(body.password),
        plan_id=1,
    )
    db.add(user)
    await db.flush()

    user_id = str(user.id)
    return TokenResponse(
        access_token=create_access_token(user_id, settings),
        refresh_token=create_refresh_token(user_id, settings),
    )


@router.post("/login", response_model=TokenResponse)
async def login(
    body: LoginRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    settings: Annotated[Settings, Depends(get_settings)],
):
    result = await db.execute(select(User).where(User.email == body.email))
    user = result.scalar_one_or_none()
    if user is None or not verify_password(body.password, user.password_hash):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail={"error": {"code": "INVALID_CREDENTIALS", "message": "Incorrect email or password", "details": {}}},
        )

    user_id = str(user.id)
    return TokenResponse(
        access_token=create_access_token(user_id, settings),
        refresh_token=create_refresh_token(user_id, settings),
    )


@router.post("/api-key", response_model=ApiKeyResponse)
async def create_api_key(
    db: Annotated[AsyncSession, Depends(get_db)],
    current_user: Annotated[User, Depends(get_current_user)],
):
    key = generate_api_key()
    current_user.api_key = key
    db.add(current_user)
    await db.flush()
    return ApiKeyResponse(api_key=key)


@router.get("/me", response_model=UserResponse)
async def get_me(
    current_user: Annotated[User, Depends(get_current_user)],
):
    return UserResponse(
        id=str(current_user.id),
        email=current_user.email,
        plan_id=current_user.plan_id,
        api_key=current_user.api_key,
        created_at=current_user.created_at.isoformat(),
    )

File: backend/node/src/auth/router.ts
import { FastifyInstance } from "fastify";
import { registerSchema, loginSchema, refreshSchema } from "./schemas";
import { registerUser, loginUser, refreshTokens, createApiKey, revokeApiKey } from "./service";
import { authenticate } from "../middleware/auth";

export async function authRouter(app: FastifyInstance) {
  app.post("/auth/register", async (request, reply) => {
    const body = registerSchema.parse(request.body);
    const result = await registerUser(app, body);
    return reply.status(201).send(result);
  });

  app.post("/auth/login", async (request, reply) => {
    const body = loginSchema.parse(request.body);
    const result = await loginUser(app, body);
    return reply.send(result);
  });

  app.post("/auth/refresh", async (request, reply) => {
    const body = refreshSchema.parse(request.body);
    const result = await refreshTokens(app, body.refreshToken);
    return reply.send(result);
  });

  app.post(
    "/auth/api-key",
    { preHandler: [authenticate] },
    async (request, reply) => {
      const userId = (request as any).userId as string;
      const result = await createApiKey(userId);
      return reply.status(201).send(result);
    }
  );

  app.delete(
    "/auth/api-key",
    { preHandler: [authenticate] },
    async (request, reply) => {
      const userId = (request as any).userId as string;
      await revokeApiKey(userId);
      return reply.status(204).send();
    }
  );

  app.get(
    "/auth/me",
    { preHandler: [authenticate] },
    async (request, reply) => {
      const user = (request as any).user;
      return reply.send({
        id: user.id,
        email: user.email,
        plan: user.plan?.name,
        apiKey: user.apiKey ? `${user.apiKey.slice(0, 12)}...` : null,
        createdAt: user.createdAt,
      });
    }
  );
}
